File: Tree/103binaryTreeZigzagLevelOrderTraversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def zigzagLevelOrder(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        q = collections.deque([root])
        res = []
        
        while q:
            temp = []
            for _ in range(len(q)):
                p = q.popleft()
                if p:
                    temp.append(p.val)
                    q.append(p.left)
                    q.append(p.right)
            if len(res) % 2:
                temp = temp[::-1]
            if temp:
                res.append(temp)
        return res 
    
    
    

File: Tree/103BinaryTreeZigzagLevelOrderTraversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        vector<vector<int>> rev;
        if(!root) return rev;
        queue<TreeNode*> q;
        q.push(root);
        bool flag = true;
        while(!q.empty()){
            int size = q.size();
            vector<int> temp(size);
            for(int i=0; i<size; i++){
                TreeNode* node = q.front(); q.pop();
                int index = flag ? i : (size-1-i);
                temp[index] = node->val;
                if(node->left) q.push(node->left);
                if(node->right) q.push(node->right);
            }
            flag = !flag;
            rev.push_back(temp);
        }
        return rev;
    }
};